homeassistant/components/kat_bulgaria/binary_sensor.py

```diff
@@ -0,0 +1,72 @@
+"""Binary sensor platform."""
+
+from datetime import datetime, timedelta
+import logging
+
+from kat_bulgaria.obligations import KatApi, KatApiResponse
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .common import generate_entity_name
+from .const import (
+    ATTR_LAST_UPDATED,
+    BINARY_SENSOR_ENTITY_PREFIX,
+    CONF_DRIVING_LICENSE,
+    CONF_PERSON_EGN,
+    CONF_PERSON_NAME,
+    DOMAIN,
+)
+
+SCAN_INTERVAL = timedelta(minutes=20)
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the platform from config_entry."""
+
+    person_name: str = str(entry.data.get(CONF_PERSON_NAME))
+    person_egn: str = str(entry.data.get(CONF_PERSON_EGN))
+    license_number: str = str(entry.data.get(CONF_DRIVING_LICENSE))
+
+    api: KatApi = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities(
+        [KatObligationsSensor(api, person_name, person_egn, license_number)], True
+    )
+
+
+class KatObligationsSensor(BinarySensorEntity):
+    """A simple sensor."""
+
+    def __init__(self, api: KatApi, name: str, egn: str, license_number: str) -> None:
+        """Initialize the sensor."""
+
+        self.api = api
+
+        self._attr_name = generate_entity_name(name)
+        self._attr_unique_id = f"{BINARY_SENSOR_ENTITY_PREFIX}{name}"
+
+        self.egn = egn
+        self.license_number = license_number
+
+    async def async_update(self) -> None:
+        """Fetch new state data for the sensor."""
+
+        resp: KatApiResponse[bool] = await self.api.async_check_obligations(
+            self.egn, self.license_number
+        )
+
+        if resp.success:
+            self._attr_is_on = resp.data
```

Comment:
What does the response data represent?

Reply:
The response is a wrapper around a HTTP call.

The whole reason I'm doing this integration is so I can have better control over the error handling of the government website. Since it regularly throws 400's and Timeouts, I treat those as expected results and just filter them out.

So in this case, this response holds the success status of the operation (`true`/`false`), as well as a payload (`resp.data`). In this case the payload is also a boolean to simplify things.

Reply:
I don't want to rename it to anything more specific, since `resp` here is a response wrapper class that can be used to return other types of data, for example the full details of any fines you have pending - amount, date, etc.

The idea is that the library is not HA-exclusive and can be used by anyone for their projects as well, that's why I used generics.

Reply:
If the api can return more than a single data type in the api call, I'd recommend using our data update coordinator helper instead to fetch the data centrally and then let the coordinator trigger an update of the entity state. That way more than a single entity can share the same data and api call.

https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

Reply:
I saw that, but every request checks the fines for a single person, so it's not applicable here.

Every GET request sends the user's Civil Number and Driver License, so you can't group them.

Reply:
To clarify, even if I include other data in the integration in the future, it will be stored in the attributes and not as separate sensor.

It will always be 1 call per 1 person.

Reply:
Yes, but one person has more than a single data point, ie will be expressed as more than a single entity.

Reply:
No, we'll create different sensors. We want to avoid using state attributes as far as possible.

Reply:
Ok then, if I decide to add more sensors in the future I will rewrite it.
However since that the usefullness of this funcionality (fine details in HA) is questionable, I doubt it will ever be implemented. I don't think you can do much with the amount of the fine you've received.

And in any case, all details for your fine show in the API only after a policeman hands you your fine in person. At this point you're better off just checking in the official website, since you have a 2 week period to pay it.

The most important functionality here is not **what** you have, but **IF** you have anything new.
This data is a single boolean, which is best represented by a single sensor.

Reply:
Ok, I think we might be miscommunicating here because of my lousy explanations, sorry.

The current version of my library has two **different** methods for checking **if** you have fines and for getting the fine details. The responses of those methods are **not** variable, e.g. one of them always has a `bool` in the `response.data`, the other one always returns a data object.

Which means if you want to check status, you call the `async_check_obligations` method, as I did on line 62.
If you want additional statuses, you need to call `async_get_obligations`, which is **not** used at all in this integration and I do not plan to use it in the foreseeable future.

tldr: `response.data` in this case will **always** be a single boolean, which translates to a single `binary_sensor` per person.

Reply:
Are you saying the integration will never expose any other data than a single boolean value per person?

Reply:
Yes

Reply:
In my experience, that's sounds unlikely to stay true, but ok.

Reply:
Unless the government changes the way they hand the fines, it will probably stay that way.

Reply:
There is already more data available, that's my point. Someone may a want to have that available in the integration.